src/lib/y2partitioner/actions/resize_partition.rb

```diff
@@ -0,0 +1,135 @@
+      #
+      # @note An error popup is shown when the partition cannot be resized.
+      #
+      # @return [Symbol, nil]
+      def run
+        return :back unless validate
+        resize
+      end
+
+    private
+
+      # @return [Y2Storage::Partition] partition to resize
+      attr_reader :partition
+
+      # @return [Y2Storage::ResizeInfo]
+      attr_reader :resize_info
+
+      # Runs the dialog to resize the partition
+      #
+      # @return [Symbol] :finish if the dialog returns :next; dialog result otherwise.
+      def resize
+        result = Dialogs::PartitionResize.run(partition, resize_info)
+        fix_end_alignment(resize_info)
+
+        result == :next ? :finish : result
+      end
+
+      # Checks whether the resize action can be performed
+      #
+      # @see Y2Storage::ResizeInfo#resize_ok?
+      #
+      # @return [Boolean] true if the resize action can be performed; false otherwise.
+      def validate
+        return true if resize_info.resize_ok?
+
+        # TODO: Distinguish the reason why it is not possible to resize, for example:
+        # * partition used by commited LVM or MD RAID
+        # * extended partition with committed logical partitions
+
+        Yast::Popup.Error(
+          # TRANSLATORS: an error popup message
+          _("This partition cannot be resized.")
+        )
+
+        false
+      end
+
+      # After the partition's size was changed during resizing, make sure the
+      # new size meets all alignment requirements, but is still between
+      # min_size and max_size. This may change the partition's size (and
+      # region).
+      #
+      # @param resize_info [ResizeInfo]
+      def fix_end_alignment(resize_info)
+        return if partition.nil?
+
+        ptable = partition.partition_table
+        return if !ptable.require_end_alignment? || partition.end_aligned?
+
+        region = ptable.align(partition.region, Y2Storage::AlignPolicy::ALIGN_END)
+        min_blocks = (resize_info.min_size.to_i / region.block_size.to_i)
+        max_blocks = (resize_info.max_size.to_i / region.block_size.to_i)
+        grain_blocks = (ptable.align_grain.to_i / region.block_size.to_i)
+
+        partition.region = fix_region_end(region, min_blocks, max_blocks, grain_blocks)
+      end
+
+      # Make sure a region's end is between min_blocks and max_blocks. If it
+      # is not, add or subtract blocks in grain_blocks increments. All sizes
+      # are specified in that region's block size.
+      #
+      # @param region [Y2Storage::Region]
+      # @param min [Fixnum]
+      # @param max [Fixnum]
+      # @param grain [Fixnum]
+      #
+      # @return [Y2Storage::Region] adjusted region
+      def fix_region_end(region, min, max, grain)
+        if region.length < min
+          region.adjust_length(grain * ((min.to_f - region.length) / grain).ceil)
+        elsif region.length > max
+          region.adjust_length(grain * ((max.to_f - region.length) / grain).floor)
```

Comment:
Sure this works?

Let's say the length of the region is 9, the max is 8 and the grain is 3. I would expect region to be adjusted by -3, so it's still aligned to a grain of 3 and is smaller than the max.

But `(8-9 / 3).floor` is 0. So the region would not be adjusted at all. 

It's Friday afternoon, so I could be completely wrong even in the most simple calculation, but...

Reply:
I think it is correct:

```
(8.0 - 9) / 3 #=> -0.33333
-0.33333.floor #=> -1
```

Reply:
Sure it's correct. As I said, Friday afternoon. :wink: